modules/tensor_mechanics/src/materials/EigenstrainHelper.C

```diff
@@ -0,0 +1,163 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#include "EigenstrainHelper.h"
+
+#include "Assembly.h"
+#include "MooseMesh.h"
+#include "SubProblem.h"
+
+// libmesh includes
+#include "libmesh/quadrature.h"
+
+EigenstrainHelper::EigenstrainHelper(const MooseObject * moose_object, bool bnd)
+  : _eh_params(moose_object->parameters()),
+    _eh_subproblem(*_eh_params.get<SubProblem *>("_subproblem")),
+    _eh_tid(_eh_params.get<THREAD_ID>("_tid")),
+    _eh_assembly(_eh_subproblem.assembly(_eh_tid)),
+    _eh_qrule(bnd ? _eh_assembly.qRuleFace() : _eh_assembly.qRule()),
+    _eh_JxW(bnd ? _eh_assembly.JxWFace() : _eh_assembly.JxW()),
+    _eh_coord(_eh_assembly.coordTransformation()),
+    _eh_q_point(bnd ? _eh_assembly.qPointsFace() : _eh_assembly.qPoints()),
+    _eh_ncols(1 + _eh_subproblem.mesh().dimension()),
+    _eh_second_order(_eh_subproblem.mesh().hasSecondOrderElements()),
+    _eh_eigsum(),
+    _eh_A(),
+    _eh_b(6),
+    _eh_AT(),
+    _eh_ATb(_eh_ncols),
+    _eh_x(6, DenseVector<Real>(_eh_ncols)),
+    _eh_vals(6),
+    _eh_adjusted_eigenstrain()
+{
+}
+
+void
+EigenstrainHelper::applyIncrementalEigenstrain(
+    const std::vector<const MaterialProperty<RankTwoTensor> *> & eigenstrains,
+    const std::vector<const MaterialProperty<RankTwoTensor> *> & eigenstrains_old,
+    MaterialProperty<RankTwoTensor> & strain)
+{
+  if (eigenstrains.size() == 0)
+    return;
+  sumEigenstrainIncremental(eigenstrains, eigenstrains_old);
+  applyEigenstrain(strain);
+}
+
+void
+EigenstrainHelper::applyTotalEigenstrain(
+    const std::vector<const MaterialProperty<RankTwoTensor> *> & eigenstrains,
+    MaterialProperty<RankTwoTensor> & strain)
+{
+  if (eigenstrains.size() == 0)
+    return;
+  sumEigenstrainTotal(eigenstrains);
+  applyEigenstrain(strain);
+}
+
+void
+EigenstrainHelper::applyEigenstrain(MaterialProperty<RankTwoTensor> & strain)
+{
+  prepareEigenstrain();
+
+  for (unsigned qp = 0; qp < _eh_qrule->n_points(); ++qp)
+  {
+    RankTwoTensor fillme;
+    if (_eh_second_order)
+    {
+      for (unsigned i = 0; i < 6; ++i)
+      {
+        _eh_vals[i] = _eh_x[i](0);
+        for (unsigned j = 1; j < _eh_ncols; ++j)
+          _eh_vals[i] += _eh_x[i](j) * _eh_q_point[qp](j - 1);
+      }
+      _eh_adjusted_eigenstrain.fillFromInputVector(_eh_vals);
```

Comment:
Is there a total eigenstrain material property that should hold the adjusted eigenstrain value?  If not, should there be such a material property? It could be confusing for a user to try to determine what happens to an eigenstrain if the adjusted eigenstrain is hidden.

Reply:
This begs a different question: would it be possible to move all of this into a material that sums and order-reduces the eigenstrain?

Reply:
I suppose this would be possible.  Send all the raw eigenstrains to a `OrderReducedEigenstrain` `Material` and then send the resulting material property to the `ComputeStrain` class.